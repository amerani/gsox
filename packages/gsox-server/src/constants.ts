export const NOTIFICATION_TOPIC = "#";
