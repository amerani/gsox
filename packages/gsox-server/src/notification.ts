// @message('notifycommand')
// class notification {

//       @str()
//       id: string;

//       @str()
//       type: string;

//       @str()
//       timestamp: string;
// }
