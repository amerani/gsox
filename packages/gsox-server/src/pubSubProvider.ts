import { PubSub, PubSubEngine } from "graphql-subscriptions";

export const pubSub: PubSubEngine = new PubSub();
