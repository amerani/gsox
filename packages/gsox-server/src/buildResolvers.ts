import "reflect-metadata";
import { withFilter } from "graphql-subscriptions";
import { NOTIFICATION_TOPIC } from "./constants";
import { pubSub } from "./pubSubProvider";

function buildResolvers() {
      return {
            Subscription: {
                ["Notification"]: {
                      subscribe: withFilter(
                            () => pubSub.asyncIterator(NOTIFICATION_TOPIC),
                            (payload) => {
                                  if (!payload) { return false; }
                                  const notification = payload.Notification;
                                  // return notification.id === id;
                                  return true;
                            },
                      ),
                },
                Ping: {
                      subscribe: () => {
                            setInterval(() => pubSub.publish("test", {
                                Ping: { id: 0 },
                            }), 2000);
                            return pubSub.asyncIterator("test");
                      },
                },
            },
          };
}

export { buildResolvers }