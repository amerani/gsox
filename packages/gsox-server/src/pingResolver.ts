import { pubSub } from "./pubSubProvider";
import { IResolvers, IResolverObject } from "../node_modules/graphql-tools";

const pingSubscribe: IResolverObject = {
      subscribe() {
            setInterval(() => pubSub.publish("test", {
                Ping: { id: 0 },
            }), 2000);
            return pubSub.asyncIterator("test");
      },
};

const pingResolver: IResolvers<any, any> = {
      Ping: pingSubscribe
}

export default pingResolver;