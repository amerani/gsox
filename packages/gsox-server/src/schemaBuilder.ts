import { Field, Type, typeDef } from "@gsox/schema";
import { gql } from "apollo-server-express";
import { withFilter } from "graphql-subscriptions";
import { makeExecutableSchema } from "graphql-tools";
import "reflect-metadata";
import { NOTIFICATION_TOPIC } from "./constants";
import { pubSub } from "./pubSubProvider";

@Type()
class Ping {

      @Field()
      public id: string;
}

@Type()
class Notification {
      @Field(null, "Int")
      public id: number;
      @Field()
      public type: string;
      @Field("timestamp")
      public data: string;
}

const ping = typeDef(new Ping());
const not = typeDef(new Notification());

const typeDefs = gql`
      type Subscription {
            notification: Notification
            ping: Ping
      }

      ${ping}

      ${not}

      type Query {
            hello: String
      }

      type schema {
            subscription: Subscription
            query: Query
      }
`;

const id = 99;

const resolvers = {
  Subscription: {
      ["notification"]: {
            subscribe: withFilter(
                  () => pubSub.asyncIterator(NOTIFICATION_TOPIC),
                  (payload) => {
                        if (!payload) { return false; }
                        const notification = payload.notification;
                        return notification.id === id;
                  },
            ),
      },
      ping: {
            subscribe: () => {
                  setInterval(() => pubSub.publish("test", {
                      ping: { id: 0 },
                  }), 2000);
                  return pubSub.asyncIterator("test");
            },
      },
  },
};

export const schema = makeExecutableSchema({typeDefs, resolvers});
