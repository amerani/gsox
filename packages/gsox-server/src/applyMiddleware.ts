import { defaults } from "@gsox/core";
import { ApolloServer } from "apollo-server-express";
import * as bodyParser from "body-parser";
import { execute, subscribe } from "graphql";
import * as http from "http";
import "reflect-metadata";
import { SubscriptionServer } from "subscriptions-transport-ws";
import { pubSub } from "./pubSubProvider";
import { buildSchema } from "./schemaBuilder";

export function applyMiddleware(app, options) {
      const curOpt = {...defaults, ...options};
      const { host, port, routes, inject } = curOpt;

      // build schema from types
      const schema = buildSchema(inject);

      // add webhook endpoint
      app.post(routes.webhook, bodyParser.json(), (req, res) => {
            const { typeName } = req.body;
            const topic = `TOPIC_${typeName}`;
            console.log(`${topic}: ${JSON.stringify(req.body)}`);
            pubSub.publish(topic, req.body);
            return res.sendStatus(200);
      });

      // init apollo server
      const path = routes.graphql;
      const apollo = new ApolloServer({ schema });
      apollo.applyMiddleware({ app, path });

      // init subscription websocket
      const server = http.createServer(app);
      SubscriptionServer.create({
            execute, schema, subscribe,
      }, {  path, server });

      return {
            apollo,
            server,
            host,
            port,
            routes,
            listen: (listener) => server.listen({ host, port }, listener),
            close: () => server.close(),
      };
}
