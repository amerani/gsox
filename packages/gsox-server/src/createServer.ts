import { ApolloServer } from "apollo-server-express";
import * as bodyParser from "body-parser";
import { execute, subscribe } from "graphql";
import { createServer as http } from "http";
import "reflect-metadata";
import { SubscriptionServer } from "subscriptions-transport-ws";
import { NOTIFICATION_TOPIC } from "./constants";
import { pubSub } from "./pubSubProvider";
import { buildSchema } from "./schemaBuilder";
import { defaults } from "@gsox/core";

export function createServer(app, options) {
      const curOpt = {...defaults, ...options}
      const { host, port, routes, inject } = curOpt;
      const schema = buildSchema(inject);

      // add webhook endpoint
      app.post(routes.webhook, bodyParser.json(), (req, res) => {
            pubSub.publish(NOTIFICATION_TOPIC, req.body);
            return res.sendStatus(200);
      });

      // init apollo server
      const server = new ApolloServer({schema});
      server.applyMiddleware({ app });

      // setup websockets endpoint
      const ws = http(app);

      ws.listen({ host, port }, () => {
            // tslint:disable-next-line:no-console
            console.log(`ðŸš€ Server ready at http://${host}:${port}${server.graphqlPath}`);

      });
      SubscriptionServer.create({
            execute,
            schema,
            subscribe,
      }, {
            path: routes.graphql,
            server: ws,
      });
      return ws;
}
