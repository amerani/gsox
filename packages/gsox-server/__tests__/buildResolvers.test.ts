import { buildResolvers } from "../src/buildResolvers";
import { Ping } from "@gsox/schema";

test('should build resolvers', () => {
      expect(buildResolvers([Ping])).toMatchSnapshot();
})