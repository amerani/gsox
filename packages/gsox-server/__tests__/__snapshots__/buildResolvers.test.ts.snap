// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should build resolvers 1`] = `
Object {
  "Subscription": Object {
    "Ping": Object {
      "subscribe": [Function],
    },
  },
}
`;
